chore(blog): publish postmortem on recent security breach

Shurtu-gal opened the pull request:
**Description**

- Postmortem on recent attack.
- 
**Related issue(s)**
<!-- If you refer to a particular issue, provide its number, otherwise, remove this section.
For example, `Resolves #123`, `Fixes #43`, or `See also #33`. The `See also #33` option will not automatically close the issue after the PR merge. -->

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Documentation**
  * Published a detailed postmortem for the Shai‑Hulud security incident covering timeline, impact, organizational response, remediation and next steps, technical attack‑chain analysis, affected parties, lessons learned, contact information, illustrative artifacts (representative malicious snippet), images, and external references to aid understanding.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Shurtu-gal committed:
docs(blog): publish postmortem on recent security breach

Signed-off-by: Shurtu-gal <ashishpadhy1729@gmail.com>

Shurtu-gal commented:
cc: @Florence-Njeri please let me know, if something needs to be added.

Shurtu-gal committed:
chore: add cover

Signed-off-by: Shurtu-gal <ashishpadhy1729@gmail.com>

Florence-Njeri commented on markdown/blog/shai-hulud-postmortem.md:
```diff
@@ -0,0 +1,101 @@
+
+All AsyncAPI packages had malicious versions published with a pattern of two bad patch releases. Example: for package version 1.2.3, versions 1.2.4 and 1.2.5 were malicious. The malicious code was designed to collect environment variables and send them to an external server controlled by the attacker. More on the technical details can be found in [Aikido's detailed analysis](https://www.aikido.de ...
+
+This led to a spread onto other dependent packages and services, including Postman, Zapier and ENS Domains. The worm works by scanning environment variables for sensitive information and publishing them in user's public repositories. The code for worm propagation is located in `bun_environment.js` file within the malicious package versions. And the same is run through `setup_bun.js` during packag ...
```
@Shurtu-gal I think this statement, `This led to a spread onto other dependent packages and services, including Postman, Zapier and ENS Domains.`  is misleading since AsyncAPI wasn't the propagation source to the other attacks but rather Patient 0 / first victim. The statement makes it seem that the worm got propagated from AsyncAPI being compromised. Maybe you can reword this?

Shurtu-gal replied:
Sure.

Florence-Njeri commented on markdown/blog/shai-hulud-postmortem.md:
```diff
@@ -0,0 +1,101 @@
+All AsyncAPI packages had malicious versions published with a pattern of two bad patch releases. Example: for package version 1.2.3, versions 1.2.4 and 1.2.5 were malicious. The malicious code was designed to collect environment variables and send them to an external server controlled by the attacker. More on the technical details can be found in [Aikido's detailed analysis](https://www.aikido.de ...
+
+This led to a spread onto other dependent packages and services, including Postman, Zapier and ENS Domains. The worm works by scanning environment variables for sensitive information and publishing them in user's public repositories. The code for worm propagation is located in `bun_environment.js` file within the malicious package versions. And the same is run through `setup_bun.js` during packag ...
+
```
Is this the right place to link: https://github.com/asyncapi/cli/commit/2efa4dff59bc3d3cecdf897ccf178f99b115d63d?

Shurtu-gal replied:
Yeah, forgot to add that.

Florence-Njeri commented on markdown/blog/shai-hulud-postmortem.md:
````diff
@@ -0,0 +1,101 @@
+  }
+```
+
+This led to [24k compromised repositories](https://github.com/search?q=Sha1-Hulud%3A+The+second+coming&type=repositories) spawning on GitHub.
````
I am not sure about this whole paragraph. Maybe instead of focusing on the whole attack and its impact we can mostly focus on: how the investgation process went, what were the indicators of compromise, the impact this sort of attack would have on our downstream users, how we quickly remediated the attack. But lmk your thoughts @Shurtu-gal because they can read all this from the aikido blog you linked.

Shurtu-gal replied:
Our response and investigation is already there under other headings. Regarding the impact aikido has already done a very good job.

Florence-Njeri commented on markdown/blog/shai-hulud-postmortem.md:
```diff
@@ -0,0 +1,101 @@
+### Next Steps
+
+- Conduct a thorough security audit of our publishing processes and infrastructure.
+- Publish a GHSA detailing the incident and our response.
```
Not everyone knows what a GHSA is, please use the full name

Shurtu-gal replied:
Sure.

Florence-Njeri commented on markdown/blog/shai-hulud-postmortem.md:
```diff
@@ -0,0 +1,101 @@
+
+Regardless of how much we prepare, security incidents can still occur. This incident has highlighted several areas for improvement:
+
+- We no longer plan to use NPM tokens for our publishing process and were in the process of switching to the recently released [Trusted Publisher](https://docs.npmjs.com/trusted-publishers/) using OIDC authentication when this incident occurred 😢. This effectively connects our GitHub repo, our CI pipeline, and the NPM registry.
```
Avoid using abbreviations if they had not been fully defined before. Not all reader know what OIDC is and maybe we can do away with the emoji since this is a formal report

Shurtu-gal replied:
Okay.

Florence-Njeri reviewed (commented):
Initial feedback

Shurtu-gal committed:
chore: apply changes

Signed-off-by: Shurtu-gal <ashishpadhy1729@gmail.com>

Shurtu-gal committed:
Merge branch 'master' into postmortem

derberg commented on markdown/blog/shai-hulud-postmortem.md:
```diff
@@ -0,0 +1,101 @@
+  - `11:46:00 UTC` — Action taken by NPM security team to unpublish the malicious packages.
+  - `13:33:00 UTC` — Suspected leak found in [vs-asyncapi-preview](https://github.com/asyncapi/vs-asyncapi-preview), as malicious version (1.0.1) was pushed to [OpenVSX Registry](https://open-vsx.org/extension/asyncapi/asyncapi-preview) through report from [Yusuke Sugamiya](https://x.com/DNPP)
+  - `13:45:00 UTC` — Revoked OpenVSX token and reported to OpenVSX security team.
+  - `14:30:00 UTC` — OpenVSX team unpublished the malicious package version.
```
I got email from them exactly at 13:52 UTC that version was unpublished

Shurtu-gal replied:
Done.

Shurtu-gal committed:
chore: fix time

Signed-off-by: Shurtu-gal <ashishpadhy1729@gmail.com>

Shurtu-gal committed:
chore: add info about workflow problem'

Signed-off-by: Shurtu-gal <ashishpadhy1729@gmail.com>

derberg commented on markdown/blog/shai-hulud-postmortem.md:
```diff
@@ -0,0 +1,102 @@
+
+## How it happened: attack chain
+
+The investigation is still ongoing, but preliminary findings suggest that the attacker gained access to our npm and OpenVSX publishing tokens, possibly through a compromised CI/CD pipeline or a leaked token in a public repository. Once they had access, they were able to publish malicious versions of our packages without our knowledge. We don't yet know when or how the initial compromise occurred, ...
```
```suggestion
The investigation is still ongoing, but preliminary findings suggest that the attacker gained access to our npm and OpenVSX publishing tokens, possibly through a compromised CI/CD pipeline or a leaked token in a public repository. Once they had access, they were able to publish malicious versions of our packages without our knowledge. We don't yet know when or how the initial compromise occurred, but as our CI/CD pipelines have recently been audited, we feel that the token leak had occurred a long time ago as we have not rotated the NPM token.

We are also 100% confident that the tokens were not stolen from any local machine. Our OpenVSX token was configured a long time ago, stored exclusively in GitHub repository secrets, and no team member ever had a local copy or password associated with it. Additionally, the npm bot user’s password had been lost for an extended period; it actually took us four months to recover access, during which 2FA was enforced as part of the account restoration. No one on the team had access to this password at any point, further ruling out the possibility of local compromise.
```

derberg commented on markdown/blog/shai-hulud-postmortem.md:
```diff
@@ -0,0 +1,102 @@
+The investigation is still ongoing, but preliminary findings suggest that the attacker gained access to our npm and OpenVSX publishing tokens, possibly through a compromised CI/CD pipeline or a leaked token in a public repository. Once they had access, they were able to publish malicious versions of our packages without our knowledge. We don't yet know when or how the initial compromise occurred, ...
+
+## Are you affected?
+
+If you have used any of the affected packages, we recommend taking the following steps:
+1. **Audit Your Environment**: Check for any unusual activity or changes in your environment variables
+2. **Update Dependencies**: Ensure that you are using the latest, non-malicious versions of the affected packages. The best way to do this is to delete your `node_modules` folder and lock files and reinstall your dependencies as the malicious versions have been unpublished.
+3. **Rotate Credentials**: If you suspect that any sensitive information may have been compromised, rotate your credentials immediately.
+4. Review your [GitHub security log](https://github.com/settings/security-log?q=action%3Arepo.create) for suspicious repositories that were created unexpectedly.
+5. Check your ~/.bashrc or ~/.zshrc for suspicious additions like sudo shutdown -h 0.
```
```suggestion
> Malicious versions are no longer in NPM. They were permanently deleted by NPM team. You no longer can access or install them.

If you have used any of the affected packages, we recommend taking the following steps:
1. **Audit Your Environment**: Check for any unusual activity or changes in your environment variables
2. **Update Dependencies**: Ensure that you are using the latest, non-malicious versions of the affected packages. The best way to do this is to delete your `node_modules` folder and lock files and reinstall your dependencies as the malicious versions have been unpublished.
3. **Rotate Credentials**: If you suspect that any sensitive information may have been compromised, rotate your credentials immediately.
4. Review your [GitHub security log](https://github.com/settings/security-log?q=action%3Arepo.create) for suspicious repositories that were created unexpectedly.
5. Check your ~/.bashrc or ~/.zshrc for suspicious additions like sudo shutdown -h 0.
```

thulieblack committed:
add clarity and accuracy

Updated the postmortem with corrected titles, dates, and improved clarity in the incident timeline and response sections.

thulieblack committed:
Update markdown/blog/shai-hulud-postmortem.md

Co-authored-by: Lukasz Gornicki <lpgornicki@gmail.com>

thulieblack committed:
minor grammar fix

Clarified details about the attack chain and token security in the postmortem.

Shurtu-gal commented on markdown/blog/shai-hulud-postmortem.md:
```diff
@@ -0,0 +1,104 @@
+- We no longer plan to use NPM tokens for our publishing process and were in the process of switching to the recently released [Trusted Publisher](https://docs.npmjs.com/trusted-publishers/) using **OIDC (OpenID Connect)** authentication when this incident occurred. This effectively connects our GitHub repo, our CI pipeline, and the NPM registry.
+
+- Have backup maintainers with publishing rights and revoking rights for tokens to reduce single points of failure.
+- Token rotation and limited scope tokens should be enforced. Our current NPM token is 3 years old.
```
@thulieblack as the token has been revoked should I remove current and change to past tense?

thulieblack replied:
Yes @Shurtu-gal you can change to past tense

thulieblack commented:
Also not sure why the tests are failing here

Shurtu-gal committed:
chore: minor fix

Signed-off-by: Shurtu-gal <ashishpadhy1729@gmail.com>

Shurtu-gal commented:
> Also not sure why the tests are failing here

This test fails intermittently: https://github.com/asyncapi/website/blob/master/tests/index.test.ts#L42

derberg commented:
/rtm
